taipy/core/job/job.py

```diff
@@ -148,36 +148,111 @@ def creation_date(self, val):
 
         Returns:
-            Optional[float]: The duration of the job execution in seconds. If the job is not
-            completed, None is returned.
+            Optional[float]: The duration of the job execution in seconds.
+                - If the job was not run, None is returned.
+                - If the job is not finished, the execution time is the duration
+                  from the running time to the current time.
         """
-        if self._execution_started_at and self._execution_ended_at:
-            return (self._execution_ended_at - self._execution_started_at).total_seconds()
-        return None
+        if Status.RUNNING.name not in self._status_change_records:
+            return None
+
+        if self.is_finished():
+            return (self.finished_time - self._status_change_records[Status.RUNNING.name]).total_seconds()
+
+        return (datetime.now() - self._status_change_records[Status.RUNNING.name]).total_seconds()
+
+    @property
+    @_self_reload(_MANAGER_NAME)
+    def pending_duration(self) -> Optional[float]:
+        """Get the duration of the job in the pending state in seconds.
+
+        Returns:
+            Optional[float]: The duration of the job in the pending state in seconds.
+                - If the job is not running, None is returned.
+                - If the job is not pending, the pending time is the duration
+                  from the submission to the current time.
+        """
+        if Status.PENDING.name not in self._status_change_records:
+            return None
+
+        if self.is_finished() or self.is_running():
+            return (
+                self._status_change_records[Status.RUNNING.name] - self._status_change_records[Status.PENDING.name]
+            ).total_seconds()
+
+        return (datetime.now() - self._status_change_records[Status.PENDING.name]).total_seconds()
+
+    @property
+    @_self_reload(_MANAGER_NAME)
+    def blocked_duration(self) -> Optional[float]:
+        """Get the duration of the job in the blocked state in seconds.
+
+        Returns:
+            Optional[float]: The duration of the job in the blocked state in seconds.
+                - If the job is not running, None is returned.
+                - If the job is not blocked, the blocked time is the duration
+                  from the submission to the current time.
+        """
+        if Status.BLOCKED.name not in self._status_change_records:
+            return None
+
+        if self.is_finished():
+            return (self.finished_time - self._status_change_records[Status.BLOCKED.name]).total_seconds()
+        if self.is_running():
+            return (
+                self._status_change_records[Status.RUNNING.name] - self._status_change_records[Status.BLOCKED.name]
+            ).total_seconds()
+
+        return (datetime.now() - self._status_change_records[Status.BLOCKED.name]).total_seconds()
```

Comment:
Is it correct?

What if the job got blocked, unblocked(pending), running, and completed?
Your method returns the difference between the completion time and the blocked time. while we expect to return PENDING - BLOCKED.

Let me propose something like the following:
```suggestion
        if self.is_blocked():
            # Currently blocked
            return (datetime.now() - self._status_change_records[Status.BLOCKED.name]).total_seconds()
        if Status.PENDING.name in self._status_change_records:
            # has been pending at a point
            return (
                self._status_change_records[Status.PENDING.name] - self._status_change_records[Status.BLOCKED.name]
            ).total_seconds()
        if self.is_finished():
            # has never been pending but is finished
            return (self.finished_time - self._status_change_records[Status.BLOCKED.name]).total_seconds()
        # has never been pending, never been finished
        # I don't see any configuration where we should enter here. 
        # Am I correct?
        raise AppropriateError
```

Reply:
You're right. I missed the PENDING state. Let me update the logic as you suggested